qpmodel/subquery.cs

```diff
@@ -144,12 +145,66 @@ LogicNode existsToMarkJoin(LogicNode nodeA, ExistSubqueryExpr existExpr)
                 topfilter = nodeAFilter.SearchAndReplace(existExpr, ConstExpr.MakeConstBool(true));
             else
                 topfilter = nodeAFilter.SearchAndReplace(existExpr, markerFilter);
+
             nodeBFilter.DeParameter(nodeA.InclusiveTableRefs());
+
+            // find all expr contains Parameter col and move it to the toper
+            var TableRefs = nodeA.InclusiveTableRefs();
+
             topfilter = topfilter.AddAndFilter(nodeBFilter);
             LogicFilter Filter = new LogicFilter(markjoin, topfilter);
+            var notDeparameterExpr = findAndfetchParameterExpr(ref nodeA);
+            if (notDeparameterExpr.Count > 0)
+            {
+                topfilter = notDeparameterExpr.AndListToExpr();
+                Filter = new LogicFilter(Filter, topfilter);
+            }
             return Filter;
         }
 
+        List<Expr> findAndfetchParameterExpr(ref LogicNode nodeA)
```

Comment:
add a comment to this function? You can say for example nodeA is ... and this functions return a list exprs of ...

Reply:
select a1 from a where a1<=3 and exists (select b.b1 from b where b.b2=a.a1 and exists (select c.c2 from c where c.c1=b.b1)

i.e. a markjoin b markjoin c

**Before:**

```
PhysicFilter  (actual rows=0)
    Output: a.a1[0]
    Filter: ({#marker}[0] and {#marker}[0])
    -> PhysicMarkJoin Left (actual rows=0)
        Output: a.a1[0],#marker,#marker
        -> PhysicScanTable a (actual rows=0)
            Output: a.a1[0]
        -> PhysicMarkJoin Left (actual rows=0)
            Output: #marker
            Filter: c.c1[1]=b.b1[0]
            -> PhysicScanTable b (actual rows=0)
                Output: b.b1[0]
***             Filter: b.b2[1]=?a.a1[0]
            -> PhysicScanTable c (actual rows=0)
                Output: c.c1[0]
```

When the method **existsToMarkJoin** is tring to construct a *markJoin*, it will leave other filter expr including expr containning parameter(i.e. b.b2[1]=?a.a1[0]) in filter of  it's own physicScanTable.  (see line with ***)

This method **findAndfetchParameterExpr** will findand fetch parameterExpr to it's top level. And next time when **existsToMarkJoin** is called, the  expr containning parameter(i.e. b.b2[1]=?a.a1[0]) will be deparameter correctly and embbeded to the filter of markjoin  where it should be.   (see line with ***)

**After:**

```
PhysicFilter  (actual rows=2)
    Output: a.a1[0]
    Filter: {#marker}[1]
    -> PhysicMarkJoin Left (actual rows=3)
        Output: a.a1[0],#marker
***     Filter: b.b2[1]=a.a1[0]
        -> PhysicScanTable a (actual rows=3)
            Output: a.a1[0]
        -> PhysicFilter  (actual rows=3, loops=3)
            Output: b.b2[0]
            Filter: {#marker}[1]
            -> PhysicMarkJoin Left (actual rows=3, loops=3)
                Output: b.b2[0],#marker
                Filter: c.c1[2]=b.b1[1]
                -> PhysicScanTable b (actual rows=3, loops=3)
                    Output: b.b2[1],b.b1[0]
                -> PhysicScanTable c (actual rows=3, loops=9)
                    Output: c.c1[0]
```